README
A Distributed DHCP Daemon
=========================

In self managed networks with decentralized client realms a central DHCP
server is not available. DDHCPD allows every realm to have its own server;
preserving roaming features for the client, by maintaining a common
consens on DHCP leases.

The implementation is flexible, meaning we do not prevent you from doing
stupid stuff, but also allow you maximal adaption.

    Usage: ddhcp [-h] [-d|-D] [-L] [-c CLT-IFACE|-S] [-i SRV-IFACE] [-t TENTATIVE-TIMEOUT] [-B BLOCK-TIMEOUT]

    -h                     This usage information.
    -c CLT-IFACE           Interface on which requests from clients are handled
    -i SRV-IFACE           Interface on which different servers communicate
    -S                     no client interface
    -t TENTATIVE           Time in seconds from when a block is announced to be claimed until a block is claimed (default: 15)
    -N NETWORK/CIDR        Network to announce and manage blocks in
    -o CODE:LEN:P1. .. .Pn DHCP Option with code, len and #len chars in decimal
    -b BLKSIZEPOW          Block size as power of two
    -B TIMEOUT             Block claim timeout (default 60s)
    -n NEEDLESS_TIMEOUT    Time until we release needless blocks (default 300s)
    -s SPARELEASES         Amount of spare leases (max: 256)
    -L                     Deactivate learning phase (otherwise the learning phase has the same time window like the block claim timeout)
    -d                     Run in background and daemonize
    -D                     Run in foreground and log to console (default)
    -C CTRL_PATH           Path to control socket
    -H COMMAND             Hook to call on events
    -V                     Print build revision
    -v                     Increase verbosity, may be specified multiple times

Build
-----

To build a production binary of DDHCPD run.

    make clean all

To increase the default logging output additional CFLAGS can be provided:

    CFLAGS="-D LOG_LEVEL_DEFAULT=5" make clean all

Additionally setting DEBUG=1 activates debug symbols:

    DEBUG=1 CFLAGS="-D LOG_LEVEL_DEFAULT=5" make clean all

To reduce the binary size to the minimum run:

    CFLAGS="-D LOG_LEVEL_LIMIT=0" make clean all

To collect statistics build with

    CFLAGS="-D DDHCPD_STATISTICS" make clean all

Depending on the chosen variant the binary size may variate between
30KiB and 1MiB.

Running
-------

On a gluon node the command may look this:

    ddhcpd -o "3:4:10.116.254.254" -o "1:4:255.255.128.0" -o "28:4:10.116.254.255" -N 10.116.224.0/20 -b 2 -s 1 -c br-client -i bat0

We are able to register DHCP options during startup but also during runtime.
Using the ddhcpdctl binary, to e.g. add the standard gateway:

    ddhcpdctl -o "3:4:10.116.254.253"

or a dedicated timeserver

    ddhcpdctl -o "42:4:10.116.254.252"

Configured options can be removed with:

    ddhcpdctl -r "42"

Be careful with this: as expressed above we do not save you from shooting yourself into your foot.

Hook
----

If you want to react to events, you can register an executable (binary or script)
via the '-H' switch. On every event that executable will be called given a list of arguments:

  <command> <action> <ip-addr> <hwaddr>

Currently <action> is one of the following keywords:

  - `lease`
    A new lease was handed out to the client indicated.

  - `release`
    The indicated lease has been released back into the address pool.

Upcoming versions of DDHCPD may introduce further action keywords.


Network analyzing
-----------------

You can count the nodes in your network wich an active ddhcpd with

    # ddhcpdctl -b | tail -n+10 | sed -e 's/\t\+/ /g' | cut -d ' ' -f 3 | sort -u | wc -l


Testing
-------

The network-test script sets up a set of virtual interfaces
which are composed of server and client interface pairs.
Those are called srvX for daemons and cltX for clients inside of the main
network namespace and server0 and client0 in the ''daemons'' network namespaces.
All srvX interfaces are bridged in br-srv; enabling all daemons to communicate
bidirectional.

Inside of the daemon's network namespace the client0 interface has the address 10.0.0.1/20.
An IPv4 address is needed here, because the DHCP part of DDHCPD binds on this interface.
It is the same address in every daemon namespace and since cltX interfaces are not
bridged this is no problem. On the server0 interface we need no IPv4 address,
because d2d communication is handled via IPv6 multicast/unicast on the local link.

You can setup a basic DDHCPD network testing environment by calling

    # ./network-test net-init <number>

which creates a virtual network for <number> of DDHCPD instances. You
will need to have sufficient privileges to create network namespaces.
Or in short: use sudo and check the network-test script upfront.

To start the (n-1)-th instance of DDHCPD in your test environment, use

    # ./network-test srv-start <n> ./ddhcpd -t 3

The above also reduces the tentative timeout from 12 to 3 seconds to
speed up testing.

After starting multiple instances, we generate DHCP requests against specific
instances of DDHCPD by starting DHCP clients pointed to the cltX interfaces,
i.e.

    # dhclient -sf /dev/null -d -v clt0

for the first DDHCPD instance. To create multiple parallel running clients
with different mac addresses, do:

    # ./network-test clt-start <n>

To unconfigure the setup call:

    # ./network-test net-stop

A simple testing routine incorporating all the above:

    # ./network-test test

Gluon Packages
--------------

A maintained gluon integration can be found at:

    https://github.com/sargon/gluon-sargon/tree/master


CHANGELOG
ddhcpd r5 (2020-01-XX)
======================
  - Fix SIGPIPE handling
  - Timed house keeping
  - Respect learning phase
  - New hook actions
  - More error handling

ddhcpd r4 (2019-12-28)
======================

  - Free unneeded blocks after some waiting time
  - Send dhcp unicast if client is renewing
  - Add dhcp inform message support 
  - Various bugfixes and small improvements

ddhcpd r3 (2019-03-07)
======================

 - Rewrite reclaim mechansim

ddhcpd r2 (2019-01-07)
======================

 - Change inquire triggering from spare blocks to spare leases.
 - Add statistics (optional activated on compile time)
 - Add gitlab-ci for debian package.
 - Major memory allocation review
 - Major type-safety review
 - Various bugfixes and small improvements

ddhcpd r1 (2018-10-01)
======================

 - Initial release


main.c
#include <arpa/inet.h>
#include <assert.h>
#include <getopt.h>
#include <math.h>
#include <netinet/in.h>
#include <sys/epoll.h>
#include <sys/un.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <signal.h>
#include <stdio.h>
#include <unistd.h>
#include <netdb.h>
#include <limits.h>

#include "block.h"
#include "ddhcp.h"
#include "dhcp.h"
#include "dhcp_packet.h"
#include "logger.h"
#include "netsock.h"
#include "packet.h"
#include "tools.h"
#include "dhcp_options.h"
#include "control.h"
#include "version.h"
#include "hook.h"
#include "statistics.h"

volatile int daemon_running = 0;

extern int log_level;

const int NET = 0;
const int NET_LEN = 10;

ATTR_NONNULL_ALL in_addr_storage get_in_addr(struct sockaddr* sa)
{
  struct sockaddr_in in;
  struct sockaddr_in6 in6;
  in_addr_storage in_store = { 0 };
  if (sa->sa_family == AF_INET) {
    memcpy(&in, sa, sizeof(in));
    in_store.in_addr = in.sin_addr;
  }

  if (sa->sa_family == AF_INET6) {
    memcpy(&in6, sa, sizeof(in6));
    in_store.in6_addr = in6.sin6_addr;
  }
  return in_store;
}

/**
 * House Keeping
 *
 * - Free timed-out DHCP leases.
 * - Refresh timed-out blocks.
 * + Claim new blocks if we are low on spare leases.
 * + Update our claims.
 */
ATTR_NONNULL_ALL void house_keeping(ddhcp_config* config) {
  DEBUG("house_keeping(blocks,config)\n");
  block_check_timeouts(config);

  uint32_t spare_leases = block_num_free_leases(config);
  int32_t leases_needed = (int32_t)config->spare_leases_needed - (int32_t)spare_leases;
  int32_t blocks_needed = leases_needed / config->block_size;

  if (leases_needed % config->block_size > 0) {
    blocks_needed++;
  }

  if (blocks_needed < 0 ) {
    block_drop_unused(config);
  } else {
    if (config->needless_marks != 0) {
      block_unmark_needless(config);
    }
    if (blocks_needed > 0) {
      block_claim(blocks_needed, config);
    }
  }

  block_update_claims(config);

  dhcp_packet_list_timeout(&config->dhcp_packet_cache);
  DEBUG("house_keeping(...) finish\n\n");
}

void add_fd(int efd, int fd, uint32_t events) {
  struct epoll_event event = { 0 };
  event.data.fd = fd;
  event.events = events;

  int s = epoll_ctl(efd, EPOLL_CTL_ADD, fd, &event);

  if (s == -1) {
    exit(1);   //("epoll_ctl");
  }
}

void del_fd(int efd, int fd, uint32_t events) {
  struct epoll_event event = { 0 };
  event.data.fd = fd;
  event.events = events;

  int s = epoll_ctl(efd, EPOLL_CTL_DEL, fd, &event);

  if (s < 0) {
    int errsv = errno;
    FATAL("%i", errsv);
    perror("epoll_ctl");
    exit(1);   //("epoll_ctl");
  }
}

ATTR_NONNULL_ALL uint32_t get_loop_timeout(ddhcp_config* config) {
  //Multiply by 500 to convert the timeout value given in seconds
  //into milliseconds AND dividing the value by two at the same time.
  //The integer overflow occuring for timeouts greater than 99.4 days is ignored here.
  return config->tentative_timeout * 500u;
}

typedef void (*sighandler_t)(int);

static sighandler_t
handle_signal(int sig_nr, sighandler_t signalhandler) {
  struct sigaction new_sig, old_sig;
  new_sig.sa_handler = signalhandler;
  sigemptyset(&new_sig.sa_mask);
  new_sig.sa_flags = SA_RESTART;

  if (sigaction(sig_nr, &new_sig, &old_sig) < 0) {
    return SIG_ERR;
  }

  return old_sig.sa_handler;
}

void handle_signal_terminate(int sig_nr) {
  if (SIGINT == sig_nr) {
    daemon_running = 0;
  } else if (SIGTERM == sig_nr) {
    daemon_running = 0;
  }
}

int main(int argc, char** argv) {

  srand((unsigned int)time(NULL));

  ddhcp_config config;
  config.block_size = 32;
  config.claiming_blocks_amount = 0;

  inet_aton("10.0.0.0", &config.prefix);
  config.prefix_len = 24;
  config.spare_leases_needed = 2;
  config.block_timeout = 60;
  config.block_refresh_factor = 4;
  config.block_needless_timeout = 300;
  config.tentative_timeout = 15;
  config.control_path = (char*)"/tmp/ddhcpd_ctl";
  config.disable_dhcp = 0;

  config.hook_command = NULL;

#ifdef DDHCPD_STATISTICS
  memset(config.statistics, 0, sizeof(long int) * STAT_NUM_OF_FIELDS);
#endif

  // DHCP
  config.dhcp_port = 67;
  INIT_LIST_HEAD(&config.options);

  INIT_LIST_HEAD(&config.claiming_blocks);

  INIT_LIST_HEAD(&config.dhcp_packet_cache);

  char* interface = (char*)"server0";
  char* interface_client = (char*)"client0";

  daemon_running = 2;

  int c;
  int show_usage = 0;
  int learning_phase = 1;

  while ((c = getopt(argc, argv, "C:c:i:St:dvVDhLb:B:N:o:s:H:n:")) != -1) {
    switch (c) {
    case 'i':
      interface = optarg;
      break;

    case 'c':
      interface_client = optarg;
      break;

    case 'b':
      config.block_size = (uint8_t)(1 << atoi(optarg));
      break;

    case 'B':
      {
        unsigned long block_timeout = strtoul(optarg, NULL, 0);
        if(!block_timeout) {
          ERROR("Block timeout must be > 0\n");
          exit(1);
        }
        if(block_timeout == ULONG_MAX && errno) {
          ERROR("Failed to parse block timeout: %s(%d)\n", strerror(errno), errno);
          exit(1);
        }
        config.block_timeout = (uint16_t)block_timeout;
      }
      break;

    case 't':
      config.tentative_timeout = (uint16_t)atoi(optarg);
      if (config.tentative_timeout < 2) {
        ERROR("Tentative timeout must at least two\n");
        exit(1);
      }
      break;

    case 'd':
      daemon_running = 1;
      break;

    case 'D':
      //We pretend we are normally running, just in another mode
      daemon_running = 2;
      break;

    case 'h':
      show_usage = 1;
      break;

    case 'L':
      learning_phase = 0;
      break;

    case 'V':
      printf("Revision: %s\n", REVISION);
      return 0;

    case 'N':
      do {
        // TODO Split prefix and cidr
        size_t optlen = strlen(optarg);
        char* cidr = strchr(optarg, '/');

        if (!cidr) {
          ERROR("Malformed network '%s'\n", optarg);
          exit(1);
        }

        if (cidr == optarg + optlen - 1) {
          ERROR("Malformed network '%s'\n", optarg);
          exit(1);
        }

        cidr[0] = '\0';
        cidr++;
        inet_aton(optarg, &config.prefix);
        config.prefix_len = (uint8_t)atoi(cidr);

        if (config.prefix_len < 8) {
          ERROR("Are you the internet? CIDR less than 8 seems strange.\n");
          exit(1);
        }
      } while (0);
      break;

    case 'S':
      config.disable_dhcp = 1;
      break;

    case 'o':
      do {
        dhcp_option* option = parse_option();
        set_option_in_store(&config.options, option);
      } while (0);
      break;

    case 's':
      config.spare_leases_needed = (uint8_t)atoi(optarg);
#if LOG_LEVEL_LIMIT >= LOG_WARNING

      if (config.spare_leases_needed == 0) {
        WARNING("This deamon will only serve roamed clients with a spare limit of zero.");
      }

#endif
      break;

    case 'C':
      config.control_path = optarg;
      break;

    case 'H':
      config.hook_command = optarg;
      break;

    case 'v':
      if (log_level < LOG_LEVEL_MAX) {
        log_level++;
      }
      break;

    case 'n':
      config.block_needless_timeout = (uint16_t)(atoi(optarg));
      break;

    default:
      printf("ARGC: %i\n", argc);
      show_usage = 1;
      break;
    }
  }

  if (show_usage) {
    printf("Usage: %s [-h] [-d|-D] [-L] [-c CLT-IFACE|-S] [-i SRV-IFACE] [-t TENTATIVE-TIMEOUT] [-B BLOCK-TIMEOUT]\n", argv[0]);
    printf("\n");
    printf("-h                     This usage information.\n");
    printf("-c CLT-IFACE           Interface on which requests from clients are handled\n");
    printf("-i SRV-IFACE           Interface on which different servers communicate\n");
    printf("-S                     no Client interface\n");
    printf("-t TENTATIVE           Time required for a block to be claimed\n");
    printf("-N NETWORK/CIDR        Network to announce and manage blocks in\n");
    printf("-o CODE:LEN:P1. .. .Pn DHCP Option with code,len and #len chars in decimal\n");
    printf("-b BLKSIZEPOW          Power over two of block size\n");
    printf("-B TIMEOUT             Block claim timeout\n");
    printf("-n NEEDLESS_TIMEOUT    Time until we release needless blocks\n");
    printf("-s SPARELEASES         Amount of spare leases (max: 256)\n");
    printf("-L                     Deactivate learning phase\n");
    printf("-d                     Run in background and daemonize\n");
    printf("-D                     Run in foreground and log to console (default)\n");
    printf("-C CTRL_PATH           Path to control socket\n");
    printf("-H COMMAND             Hook to call on events\n");
    printf("-V                     Print build revision\n");
    printf("-v                     Increase verbosity, can be specified multiple times\n");
    exit(0);
  }

  if (log_level > LOG_LEVEL_LIMIT) {
    LOG("WARNING: Requested verbosity is higher than maximum supported by this build\n");
  }

  config.number_of_blocks = (uint32_t)pow(2u, (32u - config.prefix_len - ceil(log2(config.block_size))));

  if (config.disable_dhcp) {
    config.spare_leases_needed = 0;
  }

  INFO("CONFIG: network=%s/%i\n", inet_ntoa(config.prefix), config.prefix_len);
  INFO("CONFIG: block_size=%i\n", config.block_size);
  INFO("CONFIG: #blocks=%i\n", config.number_of_blocks);
  INFO("CONFIG: #spare_leases=%i\n", config.spare_leases_needed);
  INFO("CONFIG: timeout=%i\n", config.block_timeout);
  INFO("CONFIG: refresh_factor=%i\n", config.block_refresh_factor);
  INFO("CONFIG: tentative_timeout=%i\n", config.tentative_timeout);
  INFO("CONFIG: client_interface=%s\n", interface_client);
  INFO("CONFIG: group_interface=%s\n", interface);

  //Register signal handlers
  handle_signal(SIGHUP, SIG_IGN);
  handle_signal(SIGPIPE, SIG_IGN);
  handle_signal(SIGINT, handle_signal_terminate);
  handle_signal(SIGTERM, handle_signal_terminate);

  //Daemonize if requested
  if (1 == daemon_running) {
    if (daemon(0, 0)) {
      perror("ddhcp");
      exit(1);
    }

    //Conflicts with existing logging
    //openlog("ddhcp", LOG_PID | LOG_CONS | LOG_NDELAY, LOG_DAEMON);
  }

  // init block stucture
  ddhcp_block_init(&config);

  if (dhcp_options_init(&config)) {
    FATAL("Failed to allocate memory for option store\n");
    abort();
  }

  hook_init();

  // init network and event loops
  if (netsock_init(interface, interface_client, &config) == -1) {
    return 1;
  }

  if (control_open(&config) == -1) {
    return 1;
  }

  uint8_t* buffer = (uint8_t*) malloc(sizeof(uint8_t) * 1500);

  if (!buffer) {
    FATAL("Failed to allocate network buffer\n");
    abort();
  }

  ssize_t bytes = 0;

  int efd;
  size_t maxevents = 64;
  struct epoll_event* events;

  efd = epoll_create1(0);

  if (efd == -1) {
    perror("epoll_create");
    abort();
  }

  add_fd(efd, config.mcast_socket, EPOLLIN | EPOLLET);
  add_fd(efd, config.server_socket, EPOLLIN | EPOLLET);
  add_fd(efd, config.control_socket, EPOLLIN | EPOLLET);

  if (config.disable_dhcp == 0) {
    add_fd(efd, config.client_socket, EPOLLIN | EPOLLET);
  }

  /* Buffer where events are returned */
  events = calloc(maxevents, sizeof(struct epoll_event));

  if (!events) {
    FATAL("Failed to allocate event buffer\n");
    abort();
  }

  int need_house_keeping = 0;
  uint32_t loop_timeout = config.loop_timeout = get_loop_timeout(&config);
  time_t now = time(NULL);
  // The first time we want to make housekeeping is after the learning phase, 
  // which is block_timeout long. 
  time_t timeout_time = now + config.block_timeout;

  if (!learning_phase) {
    // We want no learning phase, so reset all the timers
    loop_timeout = 0;
    timeout_time = now;
    hook(HOOK_LEARNING_PHASE_END,&config);
  }


  INFO("loop timeout: %i msecs\n", get_loop_timeout(&config));

  // TODO wait loop_timeout before first time housekeeping
  struct sockaddr_in6 sender;
  socklen_t sender_len = sizeof sender;

  do {
    int n = 0;
    do {
      n = epoll_wait(efd, events, (int)maxevents, (int)loop_timeout);
    } while (n < 0 && errno == EINTR);

    if (n < 0) {
      ERROR("epoll error (%i) %s",errno,strerror(errno));
    }

#if LOG_LEVEL_LIMIT >= LOG_DEBUG
    if (loop_timeout != config.loop_timeout) {
      DEBUG("Increase loop timeout from %i to %i\n", loop_timeout, config.loop_timeout);
    }
#endif

    loop_timeout = config.loop_timeout;

    now = time(NULL);
    if (timeout_time <= now) {
      // Our time for house keeping has come
      need_house_keeping = 1;
      // The next time for house keeping is in half the tentative timeout.
      timeout_time = now + (config.tentative_timeout >> 1);
      if (learning_phase) {
        learning_phase = 0;
        hook(HOOK_LEARNING_PHASE_END,&config);
      }
    } else { 
      need_house_keeping = 0;
    }

    for (int i = 0; i < n; i++) {
      if ((events[i].events & EPOLLERR)) {
        ERROR("Error in epoll: %i \n", errno);
        exit(1);
      } else if (config.server_socket == events[i].data.fd) {
        // DDHCP Roamed DHCP Requests
        ssize_t len;

        while ((len = recvfrom(events[i].data.fd, buffer, 1500, 0, (struct sockaddr*) &sender, &sender_len)) > 0) {
#if LOG_LEVEL_LIMIT >= LOG_DEBUG
          in_addr_storage in_addr;
          char ipv6_sender[INET6_ADDRSTRLEN];
          in_addr = get_in_addr((struct sockaddr*)&sender);
          DEBUG("Receive message from %s\n", inet_ntop(AF_INET6, &in_addr, ipv6_sender, INET6_ADDRSTRLEN));
#endif
          statistics_record(&config, STAT_DIRECT_RECV_BYTE, (long int)len);
          statistics_record(&config, STAT_DIRECT_RECV_PKG, 1);
          ddhcp_dhcp_process(buffer, len, sender, &config);
        }
      } else if (config.mcast_socket == events[i].data.fd) {
        // DDHCP Block Handling
        ssize_t len;

        while ((len = recvfrom(events[i].data.fd, buffer, 1500, 0, (struct sockaddr*) &sender, &sender_len)) > 0) {
#if LOG_LEVEL_LIMIT >= LOG_DEBUG
          in_addr_storage in_addr;
          char ipv6_sender[INET6_ADDRSTRLEN];
          in_addr = get_in_addr((struct sockaddr*)&sender);
          DEBUG("Receive message from %s\n", inet_ntop(AF_INET6, &in_addr, ipv6_sender, INET6_ADDRSTRLEN));
#endif
          statistics_record(&config, STAT_MCAST_RECV_BYTE, (long int)len);
          statistics_record(&config, STAT_MCAST_RECV_PKG, 1);
          ddhcp_block_process(buffer, len, sender, &config);
        }
        need_house_keeping = 1;
      } else if (config.client_socket == events[i].data.fd) {
        // DHCP
        ssize_t len;

        while ((len = read(config.client_socket, buffer, 1500)) > 0) {
          statistics_record(&config, STAT_DHCP_RECV_BYTE, (long int)len);
          statistics_record(&config, STAT_DHCP_RECV_PKG, 1);
          need_house_keeping |= dhcp_process(buffer, len, &config);
        }
      } else if (config.control_socket == events[i].data.fd) {
        // Handle new control socket connections
        struct sockaddr_un client_fd;
        unsigned int len = sizeof(client_fd);
        config.client_control_socket = accept(config.control_socket, (struct sockaddr*) &client_fd, &len);
        //set_nonblocking(config.client_control_socket);
        add_fd(efd, config.client_control_socket, EPOLLIN | EPOLLET);
        DEBUG("ControlSocket: new connections\n");
      } else if (events[i].events & EPOLLIN) {
        // Handle commands comming over a control_socket
        bytes = read(events[i].data.fd, buffer, 1500);

        if (handle_command(events[i].data.fd, buffer, bytes, &config) < 0) {
          ERROR("Malformed command on control socket.\n");
        }

        del_fd(efd, events[i].data.fd, 0);
        close(events[i].data.fd);
      } else if (events[i].events & EPOLLHUP) {
        del_fd(efd, events[i].data.fd, EPOLLIN);
        close(events[i].data.fd);
      }
    }

    if (need_house_keeping) {
      if (!learning_phase) {
        house_keeping(&config);
      }
    }
  } while (daemon_running);

  // TODO free dhcp_leases
  free(events);
  free(buffer);

  ddhcp_block_free(&config);

  free_option_store(&config.options);
  dhcp_packet_list_free(&config.dhcp_packet_cache);

  close(config.mcast_socket);
  close(config.client_socket);
  close(config.control_socket);

  remove(config.control_path);

  return 0;
}
